app.js

```diff
@@ -56,32 +66,45 @@ app.post('/api/shorturl', (req,res) => {
             if (!codeGenerator.checkValidShortCode(shortCode)){
                 sendError(res, "Short code must only be alphanumeric characters");
             //Make sure the short code doesn't exist in DB already
-            } else if(!checkShortCodeIsUnique(shortCode)) {
-                sendError(res, "Short code already in use");
-            } else {    
-                //Check host is valid
-                host = new URL(urlToShorten).host
-                dns.lookup(host, (err) => {
-                    if (err) {
-                        sendError(res, "Bad URL");
+            } else {
+                dbUtil.getShortCode(shortCode).then(
+                //Existing Short Code was found
+                () => {
+                    sendError(res, "Short code already in use");
+                },
+                //No single Short Code was found
+                (err) => {
+                    //No existing code returned
+                    if (err.code === 0) {
+                        //Check host is valid
+                        host = new URL(urlToShorten).host
+                        dns.lookup(host, (err) => {
+                            if (err) {
+                                sendError(res, "Bad URL");
+                            } else {
+                                //Save the association in the DB and return the short url.
+                                dbUtil.writeToDatabase(shortCode, urlToShorten, true).then(
+                                    //DB write successul
+                                    () => {
+                                        sendShortUrl(req, res, shortCode);
+                                    },
+                                    //Something went wrong
+                                    (err) => {
+                                        sendError(res, "DB insertion error: " + err.message);
```

Comment:
So essentially, you're saying if I check or not, there's a chance this will fail due to the race condition, so why both checking? I see the logic, but I think with this being such a low volume app, I'd rather guard against the chance of me generating a duplicate code over the race condition.